### PR TITLE
Correct transpose input-ordering interpolation

### PR DESCRIPTION
**Pull request by ReubenHill**
We need to ensure the target vec is zeroed for SF Reduce to
represent multTranspose in case the interpolation matrix is not
square (in which case it will have columns which are zero). This
happens when we interpolate from an input-ordering vertex-only
mesh to an immersed vertex-only mesh where the input ordering
contains points that are not in the immersed mesh. The resulting
interpolation matrix will have columns of zeros for the points
that are not in the immersed mesh. The adjoint interpolation
matrix will then have rows of zeros for those points.

### CODE REVIEW

File: firedrake/interpolation.py [hunk 1]
```diff
@@ -664,10 +664,16 @@ def interpolate(self, *function, output=None, transpose=False, **kwargs):
             if transpose:
                 mul = assembled_interpolator.handle.multTranspose
                 V = self.arguments[0].function_space()
+                result = output or firedrake.Function(V)
```
Comment (ReubenHill) [inline firedrake/interpolation.py:667]: This should be a `Cofunction` but it's not enforced anywhere else in interpolation...

File: firedrake/interpolation.py [hunk 2]
```diff
@@ -664,10 +664,16 @@ def interpolate(self, *function, output=None, transpose=False, **kwargs):
...
+                # need to zero the output in case our interpolation matrix is
+                # non-square (e.g. going from a vertex-only mesh to its input
+                # ordering where not all points in the input ordering are in
+                # the vertex-only mesh)
+                result.zero()
```
Comment (ReubenHill) [inline firedrake/interpolation.py:672]: `Cofunction` has no `zero()` method

File: firedrake/interpolation.py [hunk 3]
```diff
@@ -1429,4 +1429,14 @@ def multTranspose(self, source_vec, target_vec):
...
+            # We need to ensure the target vec is zeroed for SF Reduce to
+            # represent multTranspose in case the interpolation matrix is not
+            # square (in which case it will have columns which are zero). This
+            # happens when we interpolate from an input-ordering vertex-only
+            # mesh to an immersed vertex-only mesh where the input ordering
+            # contains points that are not in the immersed mesh. The resulting
+            # interpolation matrix will have columns of zeros for the points
+            # that are not in the immersed mesh. The adjoint interpolation
+            # matrix will then have rows of zeros for those points.
+            target_vec.getArray()[:] = 0.0
```
Comment (wence-) [inline firedrake/interpolation.py:1441]: `target_vec.zero()` makes this future-proof against GPU usage.

Comment (dham) [inline firedrake/interpolation.py:1441]:
```suggestion
            target_vec.zero()
```

**Review by dham — APPROVED** [review]

**Review by dham — CHANGES_REQUESTED** [review]
The test fails are caused by this line, so this is clearly not right.

**Comment by ReubenHill** [comment]
> The test fails are caused by this line, so this is clearly not right.

Looks like it worked before switching to an architecture agnostic approach. I wonder what's different?

File: firedrake/interpolation.py [hunk 4]
```diff
@@ -1429,4 +1429,14 @@ def multTranspose(self, source_vec, target_vec):
...
+            # We need to ensure the target vec is zeroed for SF Reduce to
+            # represent multTranspose in case the interpolation matrix is not
+            # square (in which case it will have columns which are zero). This
+            # happens when we interpolate from an input-ordering vertex-only
+            # mesh to an immersed vertex-only mesh where the input ordering
+            # contains points that are not in the immersed mesh. The resulting
+            # interpolation matrix will have columns of zeros for the points
+            # that are not in the immersed mesh. The adjoint interpolation
+            # matrix will then have rows of zeros for those points.
+            target_vec.zero()
```
Comment (ksagiyam) [inline firedrake/interpolation.py:1441]:
```suggestion
            target_vec.zeroEntries()
```

**Review by JDBetteridge — APPROVED** [review]
Approved in meeting

**Pull request merged**